File: 0567-permutation-in-string/0567-permutation-in-string.js
const checkInclusion = (s1, s2) => {
  const needs = new Map();
  const window = new Map();

  let left = 0;
  let right = 0;
  let match = 0;
  let contains = false;

  for (const char of s1) {
    const count = needs.get(char) ?? 0;
    needs.set(char, count + 1);
  }

  while (right < s2.length) {
    if (needs.has(s2[right])) {
      const count = window.get(s2[right]) ?? 0;
      window.set(s2[right], count + 1);

      if (window.get(s2[right]) === needs.get(s2[right])) {
        match++;
      }
    }

    while (match === needs.size) {
      if (right - left + 1 === s1.length) {
        contains = true;
      }

      const count = window.get(s2[left]);
      window.set(s2[left], count - 1);
      
      if (window.get(s2[left]) < needs.get(s2[left])) {
        match--;
      }

      left++;
    }

    right++;
  }

  return contains;
};

File: 0567-permutation-in-string/0567-permutation-in-string.py
class Solution:
    def checkInclusion(self, s1: str, s2: str) -> bool:
        needs = {}
        window = {}
        match = 0
        contains = False
        left = 0
        right = 0

        for char in s1:
            needs[char] = needs.get(char, 0) + 1

        while right < len(s2):
            right_char = s2[right]
            if right_char in needs:
                window[right_char] = window.get(right_char, 0) + 1
                if window[right_char] == needs[right_char]:
                    match += 1

            while match == len(needs):
                if right - left + 1 == len(s1):
                    contains = True
                    break

                left_char = s2[left]
                if left_char in needs:
                    window[left_char] -= 1
                    if window[left_char] < needs[left_char]:
                        match -= 1

                left += 1

            right += 1

        return contains
